Push Notification Implementation Guide (#1930)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/ios/push_notifications/implementation_guide.md

1. Initialize a dictionary with event metadata
2. Initialize userDefaults to retrieve and store the event data
3. If there is an existing array, append new data to the existing array and save
4. If there is not an existing array, save the new array to userDefaults

{% subtabs local %}
{% subtab Swift %}
``` swift 
func saveCustomEvent(with properties: [String: Any]? = nil) {
  // 1
  let customEventDictionary = Dictionary(eventName: "YOUR-EVENT-NAME", properties: properties)
  
  // 2
  let remoteStorage = RemoteStorage(storageType: .suite)
  
  // 3   
  if var pendingEvents = remoteStorage.retrieve(forKey: .pendingCustomEvents) as? [[String: Any]] {
    pendingEvents.append(contentsOf: [customEventDictionary])
    remoteStorage.store(pendingEvents, forKey: .pendingCustomEvents)
  } else {
  // 4
    remoteStorage.store([customEventDictionary], forKey: .pendingCustomEvents)
  }
}
```
{% endsubtab %}
{% subtab Objective-C %}
```objc
- (void)saveCustomEvent:(NSDictionary<NSString *, id> *)properties {
  // 1 

>   NSDictionary<NSString *, id> *customEventDictionary = [[NSDictionary alloc] initWithEventName:@"YOUR-EVENT-NAME" properties:properties];

Reviewer: This is a good callout. I created a category in Objective-C to add another initialization to NSDictionary. Apologies for not including it, here's the code

```
#import "NSDictionary+Helper.h"

@implementation NSDictionary (Helper)

- (id)initWithEventName:(NSString *)eventName properties:(NSDictionary *)properties {
  self = [self init];
  if (self) {
    [self setValue:eventName forKey:@"event_name"];
    
    for(id key in properties) {
      [self setValue:[properties objectForKey:key] forKey:key];
    }
  }
  return self;
}

@end
```

if this code is good we can include in the helper files?

>   NSDictionary<NSString *, id> *customEventDictionary = [[NSDictionary alloc] initWithEventName:@"YOUR-EVENT-NAME" properties:properties];

Author: @hokstuff Hey dHok, any thoughts on a third helper file like Justin suggested?

>   NSDictionary<NSString *, id> *customEventDictionary = [[NSDictionary alloc] initWithEventName:@"YOUR-EVENT-NAME" properties:properties];

Reviewer 2: Ah I see, it looks like the dictionary would have all of the information in the same level, like this:
```
{
	"event_name" : "myevent"
	"name"		 : "daniel"
	"company"    : "braze"
	// etc
}
```

If that's intended, then it looks good to me. The only comment is to update the syntax to be the newer style and use:
```
dict[@"event_name"] = eventName`
[...]
dict[key] = value
```

          eventName = eventNameValue
        } else {
          print("Invalid type for event_name key")
        }
      } else {
  // 4 
        properties[key] = value
      }
    }
  // 5    
    if let eventName = eventName {
      logCustomEvent(eventName, withProperties: properties)
    }
  }

  // 6    
  remoteStorage.removeObject(forKey: .pendingCustomEvents)
}
```
{% endsubtab %}
{% subtab Objective-C %}
```objc
- (void)logPendingEventsIfNecessary {
  RemoteStorage *remoteStorage = [[RemoteStorage alloc] initWithStorageType:StorageTypeSuite];
  NSArray *pendingEvents = [remoteStorage retrieveForKey:RemoteStorageKeyPendingCustomEvents];
  
  // 1 
  for (NSDictionary<NSString *, id> *event in pendingEvents) {
    NSString *eventName = nil;
    NSMutableDictionary *properties = [NSMutableDictionary dictionary];
    
  // 2 
    for(NSString* key in event) {
      if ([key isEqual: @"event_name"]) {
  // 3       
        if ([[event objectForKey:key] isKindOfClass:[NSString class]]) {
          eventName = [event objectForKey:key];
        } else {
          NSLog(@"Invalid type for event_name key");
        }
      } else {
  // 4 

>         [properties setValue:[event objectForKey:key] forKey:key];

Reviewer 2: Basically for each `objectForKey` call you use, it can be replaced by `dictionary[value]`. I understand we have some legacy code in our SDK still using the old syntax though

(However `dictionary[value][nestedValue]` is risky bc it can crash in the nested value call)

https://nshipster.com/object-subscripting/

Reviewer: So we can replace `[properties setValue:[event objectForKey:key] forKey:key] `with `[properties setValue:event[key] forKey:key]`?

That reads a lot better

Reviewer 3: You can simplify further:
```objc
[properties setValue:[event objectForKey:key] forKey:key];
// ==
properties[key] = event[key];
```

1. Initialize an encoded UserAttribute object with the corresponding type
2. Initialize userDefaults to retrieve and store the event data
3. If there is an existing array, append new data to the existing array and save
4. If there is not an existing array, save the new array to userDefaults

{% subtabs local %}
{% subtab Swift %}
``` swift 
func saveUserAttribute() {
  // 1 
  guard let data = try? PropertyListEncoder().encode(UserAttribute.userAttributeType("USER-ATTRIBUTE-VALUE")) else { return }
  
  // 2       
  let remoteStorage = RemoteStorage(storageType: .suite)
  
  // 3    
  if var pendingAttributes = remoteStorage.retrieve(forKey: .pendingUserAttributes) as? [Data] {
    pendingAttributes.append(contentsOf: [data])
    remoteStorage.store(pendingAttributes, forKey: .pendingUserAttributes)
  } else {
  // 4 
    remoteStorage.store([data], forKey: .pendingUserAttributes)
  }
}
```
{% endsubtab %}
{% subtab Objective-C %}
```objc
- (void)saveUserAttribute {
  // 1 
  UserAttribute *userAttribute = [[UserAttribute alloc] initWithUserField:@"USER-ATTRIBUTE-VALUE" attributeType:UserAttributeTypeEmail];
   
  NSError *error;

>   NSData *data = [NSKeyedArchiver archivedDataWithRootObject:userAttribute requiringSecureCoding:true error:&error];

Reviewer 2: nit: just ObjC style, functionality is equivalent
```suggestion
  NSData *data = [NSKeyedArchiver archivedDataWithRootObject:userAttribute requiringSecureCoding:YES error:&error];
```

>   

Reviewer 2: Add a line saying 
```
if (error != nil) {
	// log error
}
```

Same thing with code snippet below

}
 
- (void)resetStorageKeys {
  [[self defaults] removeObjectForKey:[self rawValueForKey:RemoteStorageKeyPendingCustomEvents]];
  [[self defaults] removeObjectForKey:[self rawValueForKey:RemoteStorageKeyPendingCustomAttributes]];
  [[self defaults] removeObjectForKey:[self rawValueForKey:RemoteStorageKeyPendingUserAttributes]];
}
 
- (NSUserDefaults *)defaults {
  if (!self.defaults) {
    switch (self.storageType) {
      case StorageTypeStandard:
        return [NSUserDefaults standardUserDefaults];
        break;
      case StorageTypeSuite:
        return [[NSUserDefaults alloc] initWithSuiteName:@"YOUR-SUITE-NAME"];
    }
  } else {
    return self.defaults;
  }
}
 
- (NSString*)rawValueForKey:(RemoteStorageKey)remoteStorageKey {
    switch(remoteStorageKey) {
    case RemoteStorageKeyPendingCustomEvents:
      return @"pending_custom_events";
    case RemoteStorageKeyPendingCustomAttributes:
      return @"pending_custom_attributes";
    case RemoteStorageKeyPendingUserAttributes:
      return @"pending_user_attributes";
    default:
      [NSException raise:NSGenericException format:@"Unexpected FormatType."];
  }
}
```
{% endsubtab %}
{% endsubtabs %}
{% enddetails %}

> {% details UserAttribute Helper File %}

Reviewer 2: There's 2 arrow tabs in this section in the docs deployment - doesn't look right to me

---

# Push Notification Implementation Guide

> This implementation guide covers how to leverage push notification content extensions to get the most out of your push messages. Included are three use cases built by our team, accompanying code snippets, and guidance on logging analytics. Visit our Braze Demo Repository [here](https://github.com/braze-inc/braze-growth-shares-ios-demo-app)! Please note that this implementation guide is centered around a Swift implementation, but Objective-C snippets are provided for those interested.

## Notification Content Extensions

![Push Content Extension][1]{: style="max-width:65%;border:0;margin-top:10px"}

Push notifications while seemingly standard across different platforms, offer immense customization options past what is normally implemented in the default UI. When a push notification is expanded, content notification extensions enable a custom view of the expanded push notification. 

> Push notifications can be expanded in three different ways: <br>- A long press on the push banner<br>- Swiping down on the push banner<br>- Swiping the banner to the right and selecting "View" 

Reviewer 3: ```suggestion
Push notifications can be expanded in three different ways: <br>- A long press on the push banner<br>- Swiping down on the push banner<br>- Swiping the banner to the left and selecting "View" 
```
Swiping to the right only shows the `Open` option for me.

- [Personalized Push Notifications](#personalized-push-notifications)
- [Information Capture Push Notifications](#information-capture-push-notification)

### Interactive Push Notification

Push notifications can respond to user actions inside a content extension. As of iOS 12, content extensions now have the option of being interactive! This interactivity offers many possibilities to get your users engaged in your push notification and your service. This can be achieved by having your users expand the push notification they receive.

![Push Content Extension][12]{: style="border:0"}

#### Dashboard Configuration

To set up a custom view in the dashboard you must notify Braze you would like to use a custom category, register the category, and then toggle notification buttons on. While notifications buttons aren't directly used, they are required to signal which view to display in the content extension. The pre-registered custom iOS category you provide is then checked against the code. 

{% alert tip %}
Since pushes with content extensions aren't always apparent, it is recommended to include a call to action to push your users to expand their push notifications.
{% endalert %}

> In the code, the attribute `UNNotificationExtensionCategory` is set as a string "match_game". The value given here must match what is set in the Braze dashboard. 

Reviewer 3: ```suggestion
In the content extension's _Info.plist_ file, the attribute `UNNotificationExtensionCategory` is set as a string "match_game". The value given here must match what is set in the Braze dashboard. 
```

![Personalized Push Dashboard Example][5]{: style="max-width:60%;"}

#### Handling Key-Value Pairs

The method below, `didReceive` is called when the content extension has received a notification. The key-value pairs provided in the dashboard are represented in the code through the use of a `userInfo` dictionary. This information is then parsed and sent to the push notification.

__Parsing Key-Value Pairs from Push Notifications__<br>

{% tabs %}
{% tab Swift %}
``` swift 
func didReceive(_ notification: UNNotification) {
  let userInfo = notification.request.content.userInfo
     
  guard let value = userInfo["YOUR-KEY-VALUE-PAIR"] as? String,
        let otherValue = userInfo["YOUR-OTHER-KEY-VALUE-PAIR"] as? String,
  else { fatalError("Key-Value Pairs are incorrect.")}
 
  ...
}
```
{% endtab %}
{% tab Objective-C %}
```objc
- (void)didReceiveNotification:(nonnull UNNotification *)notification {
  NSDictionary *userInfo = notification.request.content.userInfo;
   

>   if ([userInfo objectForKey:@"YOUR-KEY-VALUE-PAIR"] && [userInfo objectForKey:@"YOUR-OTHER-KEY-VALUE-PAIR"]) {

Reviewer 3: ```suggestion
  if (userInfo[@"YOUR-KEY-VALUE-PAIR"] && userInfo[@"YOUR-OTHER-KEY-VALUE-PAIR"]) {
```
Favor using modern ObjC syntax

  
  // 2       
  let remoteStorage = RemoteStorage(storageType: .suite)
  
  // 3    
  if var pendingAttributes = remoteStorage.retrieve(forKey: .pendingUserAttributes) as? [Data] {
    pendingAttributes.append(contentsOf: [data])
    remoteStorage.store(pendingAttributes, forKey: .pendingUserAttributes)
  } else {
  // 4 
    remoteStorage.store([data], forKey: .pendingUserAttributes)
  }
}
```
{% endsubtab %}
{% subtab Objective-C %}
```objc
- (void)saveUserAttribute {
  // 1 
  UserAttribute *userAttribute = [[UserAttribute alloc] initWithUserField:@"USER-ATTRIBUTE-VALUE" attributeType:UserAttributeTypeEmail];
   
  NSError *error;
  NSData *data = [NSKeyedArchiver archivedDataWithRootObject:userAttribute requiringSecureCoding:true error:&error];
  
  // 2  
  RemoteStorage *remoteStorage = [[RemoteStorage alloc] initWithStorageType:StorageTypeSuite];
  NSMutableArray *pendingAttributes = [[remoteStorage retrieveForKey:RemoteStorageKeyPendingUserAttributes] mutableCopy];
  
  // 3 
  if (pendingAttributes) {
    [pendingAttributes addObject:data];
    [remoteStorage store:pendingAttributes forKey:RemoteStorageKeyPendingUserAttributes];
  } else {
  // 4 

>     [remoteStorage store:@[ data ] forKey:RemoteStorageKeyPendingUserAttributes];

Reviewer 3: ```suggestion
    [remoteStorage store:@[data] forKey:RemoteStorageKeyPendingUserAttributes];
```
